rmw_dds_common/src/qos.cpp

```diff
@@ -0,0 +1,377 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rmw_dds_common/qos.hpp"
+
+#include <cstdarg>
+#include <cstring>
+
+#include "rcutils/snprintf.h"
+#include "rmw/error_handling.h"
+#include "rmw/qos_profiles.h"
+#include "rmw/qos_string_conversions.h"
+
+namespace rmw_dds_common
+{
+
+static bool
+operator==(rmw_time_t t1, rmw_time_t t2)
+{
+  return t1.sec == t2.sec && t1.nsec == t2.nsec;
+}
+
+static bool
+operator!=(rmw_time_t t1, rmw_time_t t2)
+{
+  return !(t1 == t2);
+}
+
+static bool
+operator<(rmw_time_t t1, rmw_time_t t2)
+{
+  if (t1.sec < t2.sec) {
+    return true;
+  } else if (t1.sec == t2.sec && t1.nsec < t2.nsec) {
+    return true;
+  }
+  return false;
+}
+
+// Returns RMW_RET_OK if successful or no buffer was provided
+// Returns RMW_RET_ERROR if there as an error copying the message to the buffer
+static rmw_ret_t
+_append_to_buffer(char * buffer, size_t buffer_size, const char * format, ...)
+{
+  // Only write if a buffer is provided
+  if (!buffer || buffer_size == 0u) {
+    return RMW_RET_OK;
+  }
+  // Determine available space left in buffer
+  size_t offset = strnlen(buffer, buffer_size);
+  size_t write_size = buffer_size - offset;
+  std::va_list args;
+  va_start(args, format);
+  int snprintf_ret = rcutils_vsnprintf(buffer + offset, write_size, format, args);
+  va_end(args);
+  if (snprintf_ret < 0) {
+    RMW_SET_ERROR_MSG("failed to append to character buffer");
+    return RMW_RET_ERROR;
+  }
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+qos_profile_check_compatible(
+  const rmw_qos_profile_t publisher_qos,
+  const rmw_qos_profile_t subscription_qos,
+  rmw_qos_compatibility_type_t * compatibility,
+  char * reason,
+  size_t reason_size)
+{
+  if (!compatibility) {
+    RMW_SET_ERROR_MSG("compatibility parameter is null");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  if (!reason && reason_size != 0u) {
+    RMW_SET_ERROR_MSG("reason parameter is null, but reason_size parameter is not zero");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  // Presume profiles are compatible until proven otherwise
+  *compatibility = RMW_QOS_COMPATIBILITY_OK;
+
+  // Initialize reason buffer
+  if (reason) {
```

Comment:
```suggestion
  if (reason && reason_size != 0u) {
```

i think there is a possibility that user can input `reason is not NULL` and `reason_size = 0`. (unlikely but possible) in that case, this overruns the reason_size. i think it is expected to do nothing if reason_size is 0.

or we could return RMW_RET_INVALID_ARGUMENT in this case, since user arguments do not make sense.

Reply:
Good catch.

We should definitely guard against zero size here. I think snprintf handles size `0` correctly (no bytes are written). I'm inclined to keep an "OK" return code, since a lot of standard library functions seem to be okay with zero lengths.

Applied your suggestion with a unit test in a8862ee